File: cpp/src/collect.cpp
#include "serial.hpp"
#include <memory>
int main(int argc,char**argv){try{
 Args a(argc,argv,{"--port","--input","--baud","--out","--bec-v"});require(a.has("--port")!=a.has("--input"),"Choose --port OR --input");
 std::unique_ptr<Serial> serial;std::ifstream file;if(a.has("--port"))serial=std::make_unique<Serial>(a.get("--port"),std::stoi(a.get("--baud","115200")));else{file.open(a.get("--input"),std::ios::binary);require(bool(file),"Cannot open input");}
 fs::path out=a.need("--out");new_run(out);std::ofstream raw((out/"raw.log").string(),std::ios::binary),csv((out/"commands.csv").string());raw.exceptions(std::ios::failbit|std::ios::badbit);csv.exceptions(std::ios::failbit|std::ios::badbit);csv<<"seq,servo,pwm,pulse_us,frequency_hz,before_tick,after_tick,rc,host_receive_ns,measured_angle_deg\n";
 Json meta={{"complete",false},{"tick_hz",nullptr},{"measurement_source","manual_video"},{"clock","host receive time is NOT actuation time"},{"options",a.values}};
 std::signal(SIGINT,on_signal);std::signal(SIGTERM,on_signal);int result=0;long long previous=-1;size_t rejected=0,gaps=0;std::string line;
 try{while(!interrupted){char c;if(serial){if(!serial->read(c))continue;}else{if(!file.get(c)){require(!file.bad(),"Input read failed");break;}}
 raw.put(c);if(c!='\n'){line+=c;require(line.size()<65536,"Unterminated serial line too long");continue;}raw.flush();
 std::smatch m;static const std::regex tick(R"(SERVO_META,tick_hz,(\d+))");if(std::regex_search(line,m,tick))meta["tick_hz"]=std::stoll(m[1]);
 std::vector<long long> v;bool valid=false;try{valid=telemetry(line,v);}catch(const std::exception&e){std::cerr<<e.what()<<'\n';rejected++;}
 if(valid){if(previous>=0&&v[0]!=previous+1){gaps++;std::cerr<<"Sequence gap/restart\n";}previous=v[0];for(auto n:v)csv<<n<<',';csv<<now_ns()<<",\n";csv.flush();}
 if(line.find("SERVO_END,")!=std::string::npos){meta["complete"]=true;meta["end_record"]=line;break;}line.clear();}}
 catch(const std::exception&e){meta["error"]=e.what();result=1;}
 if(!line.empty()&&!meta["complete"].get<bool>())meta["trailing_partial_line"]=line;
 meta["rejected_records"]=rejected;meta["sequence_gaps"]=gaps;meta["interrupted"]=bool(interrupted);write_file(out/"metadata.json",meta.dump(2));if(interrupted)std::cerr<<"Collector stopped; this does NOT stop the board.\n";return result;
 }catch(const std::exception&e){std::cerr<<e.what()<<'\n';return 1;}}

File: archive_python/collect.py
"""Read E907 console telemetry into a new run directory. Sends no commands."""
import argparse,csv,json,re,time
from pathlib import Path
PATTERN=re.compile(r'SERVO,(\d+),(\d+),(\d+),(\d+),(\d+),(\d+),(\d+),(-?\d+)')
FIELDS=['seq','servo','pwm','pulse_us','frequency_hz','before_tick','after_tick','rc']

def parse(line):
    m=PATTERN.search(line)
    if not m:return None
    r=dict(zip(FIELDS,map(int,m.groups())))
    if not 1<=r['servo']<=4 or r['pwm']!=r['servo']+3:
        raise ValueError('unexpected servo/PWM mapping')
    if not 1400<=r['pulse_us']<=1600 or r['frequency_hz']!=333:
        raise ValueError('unexpected pulse/frequency')
    return r

def main():
    p=argparse.ArgumentParser(description=__doc__)
    g=p.add_mutually_exclusive_group(required=True);g.add_argument('--port');g.add_argument('--input')
    p.add_argument('--baud',type=int,default=115200,help='must match relocated console configuration')
    p.add_argument('--out',required=True);p.add_argument('--bec-v',type=float)
    a=p.parse_args();out=Path(a.out);out.mkdir(parents=True,exist_ok=False)
    if a.port:
        import serial
        source=serial.Serial(a.port,a.baud,timeout=1)
    else:source=open(a.input,'rb')
    meta=dict(vars(a),clock='board ticks; host receive clock is NOT actuation time',
              tick_hz=None,started_unix=time.time(),measurement_source='video_manual',complete=False)
    previous=None
    try:
        with source,(out/'raw.log').open('wb') as raw,(out/'commands.csv').open('w',newline='') as csvfile:
            w=csv.DictWriter(csvfile,fieldnames=FIELDS+['host_receive_ns','measured_angle_deg']);w.writeheader()
            while True:
                line=source.readline()
                if not line:
                    if a.input:break
                    continue
                raw.write(line);raw.flush()
                text=line.decode(errors='replace')
                m=re.search(r'SERVO_META,tick_hz,(\d+)',text)
                if m:meta['tick_hz']=int(m.group(1))
                try:r=parse(text)
                except ValueError as e:print('Rejected:',e);continue
                if r:
                    if previous is not None and r['seq']!=previous+1:
                        print('WARNING sequence gap/restart',previous,r['seq'])
                    previous=r['seq'];r.update(host_receive_ns=time.monotonic_ns(),measured_angle_deg='')
                    w.writerow(r);csvfile.flush();print(text.strip())
                if 'SERVO_END,' in text:
                    meta['complete']=True;meta['end_record']=text.strip();break
    except KeyboardInterrupt:
        print('Collector stopped. This does NOT stop the board; issue servo_bench stop on its shell.')
    finally:
        (out/'metadata.json').write_text(json.dumps(meta,indent=2),encoding='utf-8')
if __name__=='__main__':main()
